### Class 7- Starter code


In [1]:
import numpy as np
import pandas as pd
from sklearn import linear_model, metrics

## Cross validation
#### Intro to cross validation with bike share data from last time. We will be modeling casual ridership. 

In [2]:
from sklearn import model_selection
wd = '../datasets/'
bikeshare = pd.read_csv(wd + 'bikeshare.csv')

In [5]:
bikeshare.tail()

instant      dteday  season  yr  mnth  hr  holiday  weekday  \
17374    17375  2012-12-31       1   1    12  19        0        1   
17375    17376  2012-12-31       1   1    12  20        0        1   
17376    17377  2012-12-31       1   1    12  21        0        1   
17377    17378  2012-12-31       1   1    12  22        0        1   
17378    17379  2012-12-31       1   1    12  23        0        1   

       workingday  weathersit  temp   atemp   hum  windspeed  casual  \
17374           1           2  0.26  0.2576  0.60     0.1642      11   
17375           1           2  0.26  0.2576  0.60     0.1642       8   
17376           1           1  0.26  0.2576  0.60     0.1642       7   
17377           1           1  0.26  0.2727  0.56     0.1343      13   
17378           1           1  0.26  0.2727  0.65     0.1343      12   

       registered  cnt  
17374         108  119  
17375          81   89  
17376          83   90  
17377          48   61  
17378          37   49

In [7]:
bikeshare.weathersit.unique()

array([1, 2, 3, 4])

#### Create dummy variables and set outcome (dependent) variable

In [4]:
weather = pd.get_dummies(bikeshare.weathersit, prefix='weather')
modeldata = bikeshare[['temp', 'hum']].join(weather[['weather_1', 'weather_2', 'weather_3']])
y = bikeshare['casual']

print(weather)

       weather_1  weather_2  weather_3  weather_4
0              1          0          0          0
1              1          0          0          0
2              1          0          0          0
3              1          0          0          0
4              1          0          0          0
5              0          1          0          0
6              1          0          0          0
7              1          0          0          0
8              1          0          0          0
9              1          0          0          0
10             1          0          0          0
11             1          0          0          0
12             1          0          0          0
13             0          1          0          0
14             0          1          0          0
15             0          1          0          0
16             0          1          0          0
17             0          1          0          0
18             0          0          1          0


**From the class ppt - Activity **

In [5]:
range_numbers = range (2, 51, 2)

for z in range_numbers:
    kf = model_selection.KFold(n_splits=z, shuffle=True)

In [6]:
mse_values = []
scores = []
n= 0
print("~~~~ CROSS VALIDATION each fold ~~~~")
for train_index, test_index in kf.split(modeldata):
    lm = linear_model.LinearRegression().fit(modeldata.iloc[train_index], y.iloc[train_index])
    mse_values.append(metrics.mean_squared_error(y.iloc[test_index], lm.predict(modeldata.iloc[test_index])))
    scores.append(lm.score(modeldata, y))
    n+=1
    print('Model', n)
    print('MSE:', mse_values[n-1])
    print('R2:', scores[n-1])


print("~~~~ SUMMARY OF CROSS VALIDATION ~~~~")
print('Mean of MSE for all folds:', np.mean(mse_values))
print('Mean of R2 for all folds:', np.mean(scores))

~~~~ CROSS VALIDATION each fold ~~~~
('Model', 1)
('MSE:', 1292.9880508220942)
('R2:', 0.31193252955063233)
('Model', 2)
('MSE:', 1674.2106432362618)
('R2:', 0.31193345261484529)
('Model', 3)
('MSE:', 1381.3757971076191)
('R2:', 0.31193174270893465)
('Model', 4)
('MSE:', 2214.8306132050629)
('R2:', 0.31192970092970562)
('Model', 5)
('MSE:', 1964.8706325731127)
('R2:', 0.31192604864138107)
('Model', 6)
('MSE:', 1696.8450182995011)
('R2:', 0.31193177319639831)
('Model', 7)
('MSE:', 1843.5834714521784)
('R2:', 0.31193205163829407)
('Model', 8)
('MSE:', 1810.3218052790501)
('R2:', 0.31193104472115463)
('Model', 9)
('MSE:', 1775.4134201604031)
('R2:', 0.31193226274703467)
('Model', 10)
('MSE:', 1880.1155266809274)
('R2:', 0.31192986613989426)
('Model', 11)
('MSE:', 1573.1110969443669)
('R2:', 0.31193392843869006)
('Model', 12)
('MSE:', 2184.2769382169186)
('R2:', 0.31193086535310632)
('Model', 13)
('MSE:', 1703.5273545839882)
('R2:', 0.3119334759612109)
('Model', 14)
('MSE:', 1781.583616555

#### Create a cross valiation with 5 folds

In [7]:
kf = model_selection.KFold(n_splits=5, shuffle=True)
# kf.get_n_splits(modeldata)

In [8]:
mse_values = []
scores = []
n= 0
print("~~~~ CROSS VALIDATION each fold ~~~~")
for train_index, test_index in kf.split(modeldata):
    lm = linear_model.LinearRegression().fit(modeldata.iloc[train_index], y.iloc[train_index])
    mse_values.append(metrics.mean_squared_error(y.iloc[test_index], lm.predict(modeldata.iloc[test_index])))
    scores.append(lm.score(modeldata, y))
    n+=1
    print('Model', n)
    print('MSE:', mse_values[n-1])
    print('R2:', scores[n-1])


print("~~~~ SUMMARY OF CROSS VALIDATION ~~~~")
print('Mean of MSE for all folds:', np.mean(mse_values))
print('Mean of R2 for all folds:', np.mean(scores))

~~~~ CROSS VALIDATION each fold ~~~~
('Model', 1)
('MSE:', 1751.7878964663466)
('R2:', 0.31185213154655067)
('Model', 2)
('MSE:', 1615.5146782245483)
('R2:', 0.31190293665184576)
('Model', 3)
('MSE:', 1533.0424543738316)
('R2:', 0.31188051762557412)
('Model', 4)
('MSE:', 1849.1569528587204)
('R2:', 0.31187436324394713)
('Model', 5)
('MSE:', 1619.2059232322558)
('R2:', 0.31189723955161242)
~~~~ SUMMARY OF CROSS VALIDATION ~~~~
('Mean of MSE for all folds:', 1673.7415810311406)
('Mean of R2 for all folds:', 0.31188143772390597)


In [6]:
# from sklearn.model_selection import KFold
# kf = KFold(n_splits=5)
# lm = linear_model.LinearRegression()
# cross_val_score(lm, modeldata, y, )

In [9]:
#Not making any folds, making a model off the entire dataset

lm = linear_model.LinearRegression().fit(modeldata, y)
print("~~~~ Single Model ~~~~")
print('MSE of single model:', metrics.mean_squared_error(y, lm.predict(modeldata)))
print('R2: ', lm.score(modeldata, y))

~~~~ Single Model ~~~~
('MSE of single model:', 1672.5811076539349)
('R2: ', 0.31193460598906442)


### Check
While the cross validated approach here generated more overall error, which of the two approaches would predict new data more accurately: the single model or the cross validated, averaged one? Why?


Answer: The cross validated, averaged one (compare MSE) 

### There are ways to improve our model with regularization. 
Let's check out the effects on MSE and R2

In [34]:
lm = linear_model.LinearRegression().fit(modeldata, y)
print("~~~ OLS ~~~")
print('OLS MSE: ', metrics.mean_squared_error(y, lm.predict(modeldata)))
print('OLS R2:', lm.score(modeldata, y))

lm = linear_model.Lasso().fit(modeldata, y)
print("~~~ Lasso ~~~")
print('Lasso MSE: ', metrics.mean_squared_error(y, lm.predict(modeldata)))
print('Lasso R2:', lm.score(modeldata, y))

lm = linear_model.Ridge().fit(modeldata, y)
print("~~~ Ridge ~~~")
print('Ridge MSE: ', metrics.mean_squared_error(y, lm.predict(modeldata)))
print('Ridge R2:', lm.score(modeldata, y))

~~~ OLS ~~~
('OLS MSE: ', 1672.5811076539349)
('OLS R2:', 0.31193460598906442)
~~~ Lasso ~~~
('Lasso MSE: ', 1725.4158160809566)
('Lasso R2:', 0.29019949592179628)
~~~ Ridge ~~~
('Ridge MSE: ', 1672.6049011306602)
('Ridge R2:', 0.31192481784314841)


### Figuring out the alphas (lambda) can be done by "hand"

In [35]:
alphas = np.logspace(-10, 10, 21)
for a in alphas:
    print('Alpha:', a)
    lm = linear_model.Ridge(alpha=a)
    lm.fit(modeldata, y)
    print(lm.coef_)
    print(metrics.mean_squared_error(y, lm.predict(modeldata)))
    
    # the resulting numbers are coefficients (as the alpha/lambda increases, the coefficients get closer to 0 )

('Alpha:', 1e-10)
[ 112.68901765  -84.01121684  -24.68489063  -21.00314493  -21.71893628]
1672.58110765
('Alpha:', 1.0000000000000001e-09)
[ 112.68901765  -84.01121684  -24.6848906   -21.00314491  -21.71893626]
1672.58110765
('Alpha:', 1e-08)
[ 112.68901765  -84.01121684  -24.6848904   -21.00314471  -21.71893606]
1672.58110765
('Alpha:', 9.9999999999999995e-08)
[ 112.68901763  -84.01121682  -24.68488837  -21.00314268  -21.71893403]
1672.58110765
('Alpha:', 9.9999999999999995e-07)
[ 112.68901745  -84.01121667  -24.68486804  -21.00312237  -21.71891373]
1672.58110765
('Alpha:', 1.0000000000000001e-05)
[ 112.68901562  -84.01121509  -24.68466472  -21.00291929  -21.71871079]
1672.58110765
('Alpha:', 0.0001)
[ 112.68899732  -84.01119938  -24.68263174  -21.00088873  -21.71668161]
1672.58110765
('Alpha:', 0.001)
[ 112.68881437  -84.01104228  -24.66232204  -20.98060316  -21.69640993]
1672.58110774
('Alpha:', 0.01)
[ 112.68698753  -84.00947323  -24.46121539  -20.77973778  -21.49568404]
1672.58111

### Or we can use grid search to make this faster

In [10]:
from sklearn.model_selection import GridSearchCV

alphas = np.logspace(-10, 10, 21)
gs = GridSearchCV(estimator=linear_model.Ridge(),
                  param_grid={'alpha': alphas}, 
                  scoring='neg_mean_squared_error')

gs.fit(modeldata, y)

GridSearchCV(cv=None, error_score='raise',
       estimator=Ridge(alpha=1.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001),
       fit_params=None, iid=True, n_jobs=1,
       param_grid={'alpha': array([  1.00000e-10,   1.00000e-09,   1.00000e-08,   1.00000e-07,
         1.00000e-06,   1.00000e-05,   1.00000e-04,   1.00000e-03,
         1.00000e-02,   1.00000e-01,   1.00000e+00,   1.00000e+01,
         1.00000e+02,   1.00000e+03,   1.00000e+04,   1.00000e+05,
         1.00000e+06,   1.00000e+07,   1.00000e+08,   1.00000e+09,
         1.00000e+10])},
       pre_dispatch='2*n_jobs', refit=True, return_train_score='warn',
       scoring='neg_mean_squared_error', verbose=0)

##### Best score 

In [11]:
print(gs.best_score_)

-1814.09369133


##### mean squared error here comes in negative, so let's make it positive.

In [12]:
print(-gs.best_score_)

1814.09369133


##### explains which grid_search setup worked best

In [37]:
print(gs.best_estimator_)

Ridge(alpha=10.0, copy_X=True, fit_intercept=True, max_iter=None,
   normalize=False, random_state=None, solver='auto', tol=0.001)


## Gradient Descent

## Demo: Application of Gradient Descent 

In [38]:
lm = linear_model.SGDRegressor()
lm.fit(modeldata, y)
print("Gradient Descent R2:", lm.score(modeldata, y))
print("Gradient Descent MSE:", metrics.mean_squared_error(y, lm.predict(modeldata)))

('Gradient Descent R2:', 0.30625282584278457)
('Gradient Descent MSE:', 1686.3926409954299)


/Users/ceceliashao/anaconda2/lib/python2.7/site-packages/sklearn/linear_model/stochastic_gradient.py:128: FutureWarning: max_iter and tol parameters have been added in <class 'sklearn.linear_model.stochastic_gradient.SGDRegressor'> in 0.19. If both are left unset, they default to max_iter=5 and tol=None. If tol is not None, max_iter defaults to max_iter=1000. From 0.21, default max_iter will be 1000, and default tol will be 1e-3.
  "and default tol will be 1e-3." % type(self), FutureWarning)


### Check: Untuned, how well did gradient descent perform compared to OLS?

Answer: 

**Gradient Descent**
('Gradient Descent R2:', 0.30625282584278457)
('Gradient Descent MSE:', 1686.3926409954299)


**-- Single Model OLS with cross-validation --**

('R2: ', 0.31193460598906442)

('MSE of single model:', 1672.5811076539349)

# Independent Practice: Bike data revisited

There are tons of ways to approach a regression problem. The regularization techniques appended to ordinary least squares optimizes the size of coefficients to best account for error. Gradient Descent also introduces learning rate (how aggressively do we solve the problem), epsilon (at what point do we say the error margin is acceptable), and iterations (when should we stop no matter what?)

For this deliverable, our goals are to:

- implement the gradient descent approach to our bike-share modeling problem,
- show how gradient descent solves and optimizes the solution,
- demonstrate the grid_search module!

While exploring the Gradient Descent regressor object, you'll build a grid search using the stochastic gradient descent estimator for the bike-share data set. Continue with either the model you evaluated last class or the simpler one from today. In particular, be sure to implement the "param_grid" in the grid search to get answers for the following questions:

- With a set of alpha values between 10^-10 and 10^-1, how does the mean squared error change?
- Based on the data, we know when to properly use l1 vs l2 regularization. By using a grid search with l1_ratios between 0 and 1 (increasing every 0.05), does that statement hold true? If not, did gradient descent have enough iterations?
- How do these results change when you alter the learning rate (eta0)?

**Bonus**: Can you see the advantages and disadvantages of using gradient descent after finishing this exercise?

# Andrew Ng video: https://www.youtube.com/watch?v=W9iWNJNFzQI

### Starter Code

In [39]:
params = {} # put your gradient descent parameters here
gs = grid_search.GridSearchCV(
    estimator=linear_model.SGDRegressor(),
    cv=cross_validation.KFold(len(modeldata), n_folds=5, shuffle=True),
    param_grid=params,
    scoring='mean_squared_error',
    )

gs.fit(modeldata, y)

print('BEST ESTIMATOR')
print(-gs.best_score_)
print(gs.best_estimator_)
print('ALL ESTIMATORS')
print(gs.grid_scores_)

NameError: name 'grid_search' is not defined

In [ ]:
## If there are concerns about the distribution of your binary target, you can undersample, 
# oversample, tinker with the `class_weight` parameter in scikit’s logistic regression, or go the tree-based route.